Compatibility with MSYS2

Would it be possible to make ssh-pageant compatible with MSYS2 environment <http://sourceforge.net/projects/msys2/>? MSYS2 is an independent rewrite of MSYS, based on modern Cygwin (POSIX compatibility layer) and MinGW-w64 with the aim of better interoperability with native Windows software. I tried compiling the ssh-pageant source code, but I got many compilation errors due to missing include files:

sys/un.h
err.h
sys/cygwin.h
sys/select.h
sys/socket.h
sys/wait.h
arpa/inet.h

It would be great to be able to use ssh-pageant together with the openssh of MSYS2.

Thank you for your quick reply. I just tried the Cygwin binaries (32-bit and 64-bit). They give me the same result as the MSYS binary:
```
$ ./ssh-pageant ssh-add -l
ssh-pageant.exe: error while loading shared libraries: ?: cannot open shared object file: No such file or directory
```
Indeed, I tried to compile ssh-pageant using MinGW-W64 gcc 4.9.2 from a shell launched via "C:\msys64\mingw64_shell.bat" and thus ran into the issue with the missing headers. I now tried using the msys/gcc version instead and the missing header errors are gone, but new errors caused by definitions in compat.h arise:
```
$ make
gcc -O2 -Werror -Wall -Wextra -MMD   -c -o main.o main.c
In file included from main.c:11:0:
compat.h:38:14: error: static declaration of ‘program_invocation_short_name’ follows non-static declaration
 static char *program_invocation_short_name = ssh_pageant_name;
              ^
In file included from /usr/include/errno.h:9:0,
                 from compat.h:24,
                 from main.c:11:
/usr/include/sys/errno.h:26:23: note: previous declaration of ‘program_invocation_short_name’ was here
 extern __IMPORT char *program_invocation_short_name;
                       ^
In file included from main.c:11:0:
compat.h:53:1: error: static declaration of ‘mkdtemp’ follows non-static declaration
 mkdtemp(char *template)
 ^
In file included from /usr/include/string.h:10:0,
                 from /usr/include/sys/un.h:14,
                 from compat.h:16,
                 from main.c:11:
/usr/include/stdlib.h:110:8: note: previous declaration of ‘mkdtemp’ was here
 char * _EXFUN(mkdtemp,(char *));
        ^
In file included from main.c:11:0:
compat.h:63:1: error: static declaration of ‘strlcpy’ follows non-static declaration
 strlcpy(char *dst, const char *src, size_t size)
 ^
In file included from /usr/include/string.h:10:0,
                 from /usr/include/sys/un.h:14,
                 from compat.h:16,
                 from main.c:11:
/usr/include/string.h:119:8: note: previous declaration of ‘strlcpy’ was here
 size_t _EXFUN(strlcpy,(char *, const char *, size_t));
        ^
In file included from main.c:11:0:
compat.h:85:1: error: static declaration of ‘accept4’ follows non-static declaration
 accept4(int sockfd, struct sockaddr *addr, socklen_t *addrlen, int flags)
 ^
In file included from compat.h:30:0,
                 from main.c:11:
/usr/include/sys/socket.h:26:7: note: previous declaration of ‘accept4’ was here
   int accept4 (int, struct sockaddr *__peer, socklen_t *, int flags);
       ^
In file included from main.c:11:0:
compat.h:104:1: error: static declaration of ‘cygwin_conv_path’ follows non-static declaration
 cygwin_conv_path (unsigned what, const void *from, void *to, size_t size)
 ^
In file included from compat.h:29:0,
                 from main.c:11:
/usr/include/sys/cygwin.h:74:16: note: previous declaration of ‘cygwin_conv_path’ was here
 extern ssize_t cygwin_conv_path (cygwin_conv_path_t what, const void *from,
                ^
In file included from main.c:11:0:
compat.h: In function ‘cygwin_conv_path’:
compat.h:111:5: error: implicit declaration of function ‘cygwin_conv_to_posix_path’ [-Werror=implicit-function-declaration]
     if (((what & CCP_RELATIVE) ? cygwin_conv_to_posix_path(from, posix)
     ^
compat.h:112:17: error: implicit declaration of function ‘cygwin_conv_to_full_posix_path’ [-Werror=implicit-function-declaration]
                 : cygwin_conv_to_full_posix_path(from, posix)) != 0)
                 ^
In file included from main.c:29:0:
winpgntc.h: At top level:
winpgntc.h:21:23: error: conflicting types for ‘uint32_t’
 typedef unsigned long uint32_t;
                       ^
In file included from /usr/lib/gcc/x86_64-pc-msys/4.9.2/include/stdint.h:9:0,
                 from /usr/include/cygwin/socket.h:19,
                 from /usr/include/sys/un.h:15,
                 from compat.h:16,
                 from main.c:11:
/usr/include/stdint.h:34:22: note: previous declaration of ‘uint32_t’ was here
 typedef unsigned int uint32_t;
                      ^
cc1: all warnings being treated as errors
<builtin>: recipe for target 'main.o' failed
make: *** [main.o] Error 1
```
MSYS2's gcc pre-defines the following Cygwin- and MSYS-related macros:
```
$ gcc   -dM -E -x c /dev/null | grep CYGWIN
#define __CYGWIN__ 1
$ gcc   -dM -E -x c /dev/null | grep MSYS
#define __MSYS__ 1
```
Unfortunately, there doesn't seem to be a dedicated __MSYS2__ or similar. If MSYS did not define __CYGWIN__, this could be a way to distinguish MSYS2 from MSYS. Unfortunately, I don't have access to an MSYS installation to test it.

Using the proposed ```defined(__MSYS__) && !defined(__NEWLIB__)``` in compat.h reduces the amount of compilation errors to:
```
$ make
gcc -O2 -Werror -Wall -Wextra -MMD   -c -o main.o main.c
In file included from main.c:29:0:
winpgntc.h:21:23: error: conflicting types for ‘uint32_t’
 typedef unsigned long uint32_t;
                       ^
In file included from /usr/lib/gcc/x86_64-pc-msys/4.9.2/include/stdint.h:9:0,
                 from /usr/include/cygwin/socket.h:19,
                 from /usr/include/sys/un.h:15,
                 from compat.h:16,
                 from main.c:11:
/usr/include/stdint.h:34:22: note: previous declaration of ‘uint32_t’ was here
 typedef unsigned int uint32_t;
                      ^
<builtin>: recipe for target 'main.o' failed
make: *** [main.o] Error 1
```

--- a/compat.h
+++ b/compat.h
@@ -18,7 +18,7 @@
 #define UNIX_PATH_MAX UNIX_PATH_LEN
 #endif
 
-#ifdef __MSYS__
+#if defined(__MSYS__) && !defined(__NEWLIB__)
 
 #include <unistd.h>
 #include <errno.h>
@@ -123,6 +123,6 @@ cygwin_conv_path (unsigned what, const void *from, void *to, size_t size)
 
 #include <err.h>
 
-#endif
+#endif // defined(__MSYS__) && !defined(__NEWLIB__)
 
 #endif /* __COMPAT_H__ */

--- a/winpgntc.h
+++ b/winpgntc.h
@@ -15,11 +15,11 @@
 
 #define AGENT_MAX_MSGLEN  8192
 
-#ifdef __MSYS__
+#if defined(__MSYS__) && !defined(__NEWLIB__)
 // MSYS doesn't have stdint.h or uint32_t,
 // but its ntohl wants unsigned long anyway.
 typedef unsigned long uint32_t;
-#endif
+#endif // defined(__MSYS__) && !defined(__NEWLIB__)
 
 extern void agent_query(void *buf);
 

